wrapper/main.cpp

```diff
@@ -0,0 +1,546 @@
+    child_args[i - optind] = argv[i];
+  }
+  // This is a nightmare of C memory management. We're just going to
+  // duplicate and leak all these strings, because I don't care anymore.
+  child_args[argc - optind] = strdup("--output-dir");
+  child_args[argc - optind + 1] = temporary_directory;
+  child_args[argc - optind + 2] = strdup("--sample-sheet");
+  child_args[argc - optind + 3] = strdup(samplesheet.c_str());
+  child_args[argc - optind + 4] = nullptr;
+
+  // In UNIX, we can't create a new process from scratch; we duplicate the
+  // existing process using fork() and the return value of fork() tells us
+  // whether we are the parent (by giving us the child PID) or the child; the
+  // child can then set itself up to be trace and execute bcl2fastq which
+  // replaces its own state with that new program
+  errno = 0;
+  const auto pid = fork();
+  if (pid < 0) {
+    // Fork failed. Not much we can do
+    perror("fork");
+    return 1;
+  } else if (pid == 0) {
+    // This is the child process. Tell the OS we consent to be traced.
+    errno = 0;
+    if (ptrace(PTRACE_TRACEME, 0, 0, 0) < 0) {
+      perror("fork");
+      return 1;
+    }
+    // Send a stop signal to ourselves so we don't do anything until the parent
+    // is ready to trace
+    kill(getpid(), SIGSTOP);
+    errno = 0;
+    // Run bcl2fastq; this will replace this program's code with the child's,
+    // but keep our ptrace settings and open files intact
+    execvp(child_args[0], child_args);
+    perror("execvp");
+    return 1;
+  } else {
+    errno = 0;
+    int status = 0;
+    // Wait for the child to stop itself
+    waitpid(pid, &status, 0);
+    if (!WIFSTOPPED(status)) {
+      std::cerr << "Child process got a signal that wasn't the STOP it was "
+                   "supposed to send to itself."
+                << std::endl;
+      return 1;
+    }
+    errno = 0;
+    // Tell the OS we are going to trace it and if we die, kill the child.
+    if (ptrace(PTRACE_SETOPTIONS, pid, 0,
+               PTRACE_O_TRACESYSGOOD | PTRACE_O_EXITKILL) < 0) {
+      perror("ptrace");
+      return 1;
+    }
+    // Keep checking on the child until it exists
+    if (!run_process(pid)) {
+      std::cerr << "Giving up." << std::endl;
+      return 1;
+    }
+  }
+  std::cerr << "Reading stats file..." << std::endl;
+  Json::Value stats_json;
+  std::ifstream stats_data(std::string(temporary_directory) +
+                           "/Stats/Stats.json");
+  stats_data >> stats_json;
+
+  std::set<int> reads;
+  for (Json::ArrayIndex i = 0; i < stats_json["ConversionResults"].size();
+       i++) {
+    auto &cr = stats_json["ConversionResults"][i];
+    for (Json::ArrayIndex j = 0; j < cr["Undetermined"]["ReadMetrics"].size();
+         j++) {
+      auto &rm = cr["Undetermined"]["ReadMetrics"][j];
+      for (Json::ArrayIndex k = 0; k < rm.size(); k++) {
+        reads.insert(rm["ReadNumber"].asInt());
+      }
+    }
+  }
+
+  std::vector<std::string> fastqs;
```

Comment:
Perhaps add one line here explaining that the fastq files are being extracted from the temporary directory?